vmms/dockerSSH.py

```diff
@@ -0,0 +1,223 @@
+        return "%s-%s-%s" % (config.Config.PREFIX, id, name)
+
+    def domainName(self, vm):
+        """ Returns the domain name that is stored in the vm
+        instance.
+        """
+        return vm.domain_name
+
+    #
+    # VMMS API functions
+    #
+    def initializeVM(self, vm):
+        """ initializeVM -  Nothing to do for initializeVM
+        """
+        return vm
+
+    def waitVM(self, vm, max_secs):
+        """ waitVM - Nothing to do for waitVM
+        """
+        return
+
+    def copyIn(self, vm, inputFiles):
+        """ copyIn - Create a directory to be mounted as a volume
+        for the docker containers. Copy input files to this directory.
+        """
+        instanceName = self.instanceName(vm.id, vm.image)
+        # Create a fresh volume
+        volume_path = config.Config.DOCKER_VOLUME_PATH + instanceName +'/'
+        os.makedirs(volume_path)
+        for file in inputFiles:
+            shutil.copy(file.localFile, volume_path + file.destFile)
+            self.log.info('Copied in file %s to %s' % (file.localFile, volume_path + file.destFile))
+        return 0
+
+    def runJob(self, vm, runTimeout, maxOutputFileSize):
+        """ runJob - Run a docker container by doing the follows:
+        - mount directory corresponding to this job to /home/autolab
+          in the container
+        - run autodriver with corresponding ulimits and timeout as
+          autolab user
+        """
+        instanceName = self.instanceName(vm.id, vm.image)
+        args = ['docker', 'run', '--name', instanceName, '-v']
+        args = args + ['%s:%s' % 
+                (config.Config.DOCKER_VOLUME_PATH + instanceName, '/home/mount')]
+        args = args + [config.Config.DOCKER_IMAGE]
+        args = args + ['sh', '-c']
+
+        autodriverCmd = 'autodriver -u %d -f %d -t %d -o %d autolab &> output/feedback' % \
+                        (config.Config.VM_ULIMIT_USER_PROC, 
+                        config.Config.VM_ULIMIT_FILE_SIZE,
+                        runTimeout, config.Config.MAX_OUTPUT_FILE_SIZE)
+
+        args = args + ['cp -r mount/* autolab/; su autolab -c "%s"; \
+                        cp output/feedback mount/feedback' % autodriverCmd]
+
+        self.log.info('Running job: %s' % str(args))
+
+        return timeout(args, runTimeout)
+
+
+    def copyOut(self, vm, destFile):
+        """ copyOut - Copy the autograder feedback from container to
+        destFile on the Tango host.
+        """
+        instanceName = self.instanceName(vm.id, vm.image)
+        volume_path = config.Config.DOCKER_VOLUME_PATH + instanceName
+        print os.listdir(volume_path)
+        print volume_path + '/feedback'
+        shutil.move(volume_path + '/feedback', destFile)
+        self.log.info('Copied feedback file to %s' % destFile)
+        
+        # Must always clean up containers in order to maintain statelessness.
+        # A solution with `docker attach` could be explored.
+        self.destroyVM(vm)
```

Comment:
Hmm. While I'm not sure that we'll need to be reusing containers, we might
be able to do something similar where if we know that the next job in the
queue will use the same auto driver and makefile we copy both in and out,
but this could be a security risk.
On Wed, Mar 25, 2015 at 11:48 AM Mihir Pandya <notifications@github.com>
wrote:

> In vmms/dockerSSH.py
> <https://github.com/autolab/Tango/pull/45#discussion_r27133786>:
>
> > +
> > +
> > +    def copyOut(self, vm, destFile):
> > +        """ copyOut - Copy the autograder feedback from container to
> > +        destFile on the Tango host.
> > +        """
> > +        instanceName = self.instanceName(vm.id, vm.image)
> > +        volume_path = config.Config.DOCKER_VOLUME_PATH + instanceName
> > +        print os.listdir(volume_path)
> > +        print volume_path + '/feedback'
> > +        shutil.move(volume_path + '/feedback', destFile)
> > +        self.log.info('Copied feedback file to %s' % destFile)
> > +
> > +        # Must always clean up containers in order to maintain statelessness.
> > +        # A solution with `docker attach` could be explored.
> > +        self.destroyVM(vm)
>
> Well, I wanted to explore the possibility of reusing a container instead
> of destroying it every time after copyOut. It is not to be actually used
> in destroyVM.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/autolab/Tango/pull/45/files#r27133786>.
>